src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -226,10 +226,34 @@ public async Task ApplyBatchChangeToWorkspaceMaybeAsync(bool useAsync, Action<So
         {
             using (useAsync ? await _gate.DisposableWaitAsync().ConfigureAwait(false) : _gate.DisposableWait())
             {
-                var solutionChanges = new SolutionChangeAccumulator(Workspace.CurrentSolution);
-                mutation(solutionChanges);
-
-                ApplyBatchChangeToWorkspace_NoLock(solutionChanges);
+                // We need the data from the accumulator across the lambda callbacks to SetCurrentSolutionAsync, so declare
+                // it here. It will be assigned in `transformation:` below (which may happen multiple times if the
+                // transformation needs to rerun).  Once the transformation succeeds and is applied, the
+                // 'onBeforeUpdate/onAfterUpdate' callbacks will be called, and can use the last assigned value in
+                // `transformation`.
+                SolutionChangeAccumulator solutionChanges = null!;
+
+                await Workspace.SetCurrentSolutionAsync(
+                    useAsync,
+                    transformation: oldSolution =>
+                    {
+                        solutionChanges = new SolutionChangeAccumulator(oldSolution);
+                        mutation(solutionChanges);
+
+                        // If the accumulator showed no changes, return the old solution.  This will ensure that
+                        // SetCurrentSolutionAsync bails out immediately and no further work is done.
+                        return solutionChanges.HasChange ? solutionChanges.Solution : oldSolution;
+                    },
+                    changeKind: (_, _) => (solutionChanges.WorkspaceChangeKind, solutionChanges.WorkspaceChangeProjectId, solutionChanges.WorkspaceChangeDocumentId),
+                    onBeforeUpdate: (_, _) =>
+                    {
+                        // Clear out mutable state not associated with the solution snapshot (for example, which documents are
+                        // currently open).
+                        foreach (var documentId in solutionChanges.DocumentIdsRemoved)
+                            Workspace.ClearDocumentData(documentId);
+                    },
+                    onAfterUpdate: null,
+                    CancellationToken.None).ConfigureAwait(false);
```

Comment:
this change at least moves this specific PSPF to use the workspace helper that makes stateless transformations against the it's current-snapshot and uses optimistic concurrency to try to apply it if nothing else intervened.  If anything else did, it will back off and continue trying until it succeeds.  Only once it does does it then update the current solution and then calls the other lambdas which can make stateful changes.  

The goal is to move *everything* in this type over to that pattern.  HOwever, it is difficult given the disparate patterns used here, as well as that this type is itself then called by other external entities making many changes and expecting them all to be atomic.  It's going to take a fair amount of work to untangle everything.

Reply:
Yep, refactoring this one method at a time is going to be the right approach.